storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java

```diff
@@ -153,6 +173,7 @@ public void run() {
 
     private void processConsumerRecord(ConsumerRecord<byte[], byte[]> record) {
         final RemoteLogMetadata remoteLogMetadata = serde.deserialize(record.value());
+        onConsume.accept(remoteLogMetadata);
```

Comment:
This is not the correct way to capture the events. Assume that the testcase don't want to process an event (`shouldProcess` check returns false). We don't want that event to be captured. 

Instead, we can have a setter method for `RemotePartitionMetadataStore` and pass a custom implementation similar to [DummyEventHandler](https://github.com/apache/kafka/blob/trunk/storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTaskTest.java#L374) where we can capture the event and delegate it to the real implementation.

Reply:
Thanks for the suggestion! This made me realise there's also a possible race where even after `RemotePartitionMetadataStore::handleRemoteLogSegmentMetadata` is invoked, the assertions on `topicBasedRlmm().listRemoteLogSegments` may fail because `remoteLogMetadataCache.isInitialized()` may return false.

Inspired by your suggestion to hook on `RemotePartitionMetadataStore`, I've modified `TopicBasedRemoteLogMetadataManagerHarness` to accept a spy object for it which is passed down to `ConsumerTask`. The tests are modified to ensure `handleRemoteLogSegmentMetadata` and `markInitialized` are invoked appropriate number of times.

Reply:
Thanks for updating the test, the approach LGTM!

Nit: Why are we using the `supplier` pattern instead of adding a setter to `TopicBasedRemoteLogMetadataManager` and marking it as `visibleForTesting`?

Reply:
> Nit: Why are we using the supplier pattern instead of adding a setter to TopicBasedRemoteLogMetadataManager and marking it as visibleForTesting?

IIUC, you're alluding to something similar we do for `remoteLogMetadataTopicPartitioner` at https://github.com/apache/kafka/blob/8a9dd2beda90f04180e71b406657d9da388d359e/storage/src/test/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManagerHarness.java#L108.

That can work, but I feel it's very easy to introduce a race inadvertently since `TopicBasedRemoteLogMetadataManager::configure` spawns a thread (https://github.com/apache/kafka/blob/8a9dd2beda90f04180e71b406657d9da388d359e/storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java#L368). In fact, `remoteLogMetadataTopicPartitioner` is prone to a race, where if the test thread yields before line 109 is executed, the `ProducerManager` and `ConsumerManager` instances can get instantiated with incorrect `remoteLogMetdataTopicPartitioner` instance (https://github.com/apache/kafka/blob/8a9dd2beda90f04180e71b406657d9da388d359e/storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/TopicBasedRemoteLogMetadataManager.java#L421).

We can avoid it by invoking the setter before calling `TopicBasedRemoteLogMetadataManager::configure` but I feel it's easier to enforce it by using a `Supplier` instead. Either way, I feel this race should be fixed as well now :)